compiler/rustc_parse/src/parser/path.rs

```diff
@@ -373,7 +374,52 @@ impl<'a> Parser<'a> {
                     .into()
                 } else {
                     // `(T, U) -> R`
-                    let (inputs, _) = self.parse_paren_comma_seq(|p| p.parse_ty())?;
+                    let prev_token_before_parsing = self.prev_token.clone();
+
+                    let mut snapshot = None;
+                    if self.may_recover()
+                        && (self.token.can_begin_expr() || self.token.can_begin_pattern())
+                    {
+                        snapshot = Some(self.create_snapshot_for_diagnostic());
+                    }
+
+                    let (inputs, _) = match self.parse_paren_comma_seq(|p| p.parse_ty()) {
+                        Ok(output) => output,
+                        Err(mut error) if prev_token_before_parsing.kind == token::ModSep => {
```

Comment:
I have moved it to a separate method. The right indentation is eye-twitching :smile: .

However, passing the reference to the diag seems unnatural. Should I rather create a new diagnostic span instead of attaching it to the error?

Reply:
I understand what you mean but actually there are a lot of functions in `compiler/` that do this (taking a `&mut Diag<'_>` and modifying it under certain conditions), so it's perfectly fine :)